text/3503-frontmatter.md

```diff
@@ -0,0 +1,579 @@
+  - Unpredictable location (both the doc comment and the cargo code block within it)
+  - Visual clutter (where clutter is overwhelming already in Rust)
+- New style of structured comment for the ecosystem to support with potential
+  compatibility issues, likely requiring a new edition
+- Assuming it can't be parsed with `syn` and either we need to write a
+  sufficiently compatible comment parser or pull in a much larger rust parser
+  to extract and update comments.
+  - Like with doc comments, this should map to an attribute and then we'd just start the MVP with that attribute
+
+### Alternative 6: Macro
+
+```rust
+#!/usr/bin/env cargo
+
+cargo! {
+[package]
+edition = "2018"
+}
+
+fn main() {
+}
+```
+Benefits
+- Parsers are available to make this work (e.g. `syn`)
+- Users can edit/copy/paste the manifest without dealing with leading characters
+
+Downsides
+- When discussing with a Rust crash course teacher, it was felt their students would have a hard time learning to write these manifests from scratch
+  - Unpredictable location (both the doc comment and the cargo code block within it)
+- The `cargo` macro would need to come from somewhere (`std`?) which means it is taking on `cargo`-specific knowledge
+  - An unexplored direction we could go with this is a `meta!` macro (e.g. we'd need to have a format marker in it)
+- A lot of tools/IDEs have problems in dealing with macros
+- Free-form rust code makes it harder for cargo to make edits to the manifest
+- Either we expose `syn`s lesser parse errors or we skip errors, deferring to rustc's, but then have the wrong behavior on commands that don't invoke rustc, like `cargo metadata`
+
+### Alternative 7: Presentation Streams
+
+```rust
+#!/usr/bin/env cargo
+
+fn main() {
+}
+
+---Cargo.toml
+[package]
+edition = "2018"
+```
+YAML allows several documents to be concatenated together variant
+[presentation streams](https://yaml.org/spec/1.2.2/#323-presentation-stream)
+which might seem familiar as this is frequently used in static-site generators
+for adding frontmatter to pages.
+What if we extended Rust's syntax to allow something similar?
+
+Benefits
+- Flexible for other content
+- Users can edit/copy/paste the manifest without dealing with leading characters
+
+Downsides
+- Difficult to parse without assistance from something like `syn` as we'd need to distinguish what the start of a stream is vs content of a string literal
+- Being a new file format (a "text tar" format), there would be a lot of details to work out, including
+  - How to delineate and label documents
+  - How to allow escaping to avoid conflicts with content in a documents
+  - Potentially an API for accessing the document from within Rust
+- Unfamiliar, new syntax, unclear how it will work out for newer users
+
+# Prior art
+[prior-art]: #prior-art
+
+See also [Single-file scripts that download their dependencies](https://dbohdan.com/scripts-with-dependencies)
+which enumerates the syntax used by different tools.
```

Comment:
(Commenting here since I can't find any other spot to comment)

I think that [Bazel's optional crate naming scheme](https://github.com/bazelbuild/rules_rust/discussions/1008) is honestly extremely interesting approach that tries to make declaring dependencies inside of a source file. It's not _exactly_ the same thing as this, but it's not that different either.